Add `Correctors` syntax

kyleeswanson opened the pull request:
## Reason
Adding syntax for Correctors 

## Impact
Allows better organization of different UserObject capabilities

Closes  #28339

Commit:
Adding Correctors folders in framework - Issue #28339

Commit:
Editing files to work with Correctors syntax - Issue #28339

Commit:
Add Moose.C changes for correctors
Created AddCorrectorAction.md - Issue #28339

dschwen commented:
This is neat. I was about to complain that we already have a user object that renormalizes solutions, but it looks like you're moving it to a new system dedicated to solution "postprocessing".

dschwen commented on framework/doc/content/source/correctors/PointwiseRenormalizeVector.md:
```diff
@@ -0,0 +1,21 @@
+
+!syntax description /Correctors/PointwiseRenormalizeVector
+
+This user object can pointwise renormalize the solution for a set of variables, taking each variable as the component of a vector and scaling the variables to obtain the user specified L2-norm.
```
Not a user object anymore

hugary1995 replied:
I have a related comment: I understand that right now correctors are basically userobjects in the backend. But should we still advertise that fact? Not just in this documentation markdown, I also saw similar emphasis in docstrings elsewhere. For example, the postprocessors also derive from userobject, but we don't explicitly emphasize that fact anywhere. Likewise, I feel we probably don't need to mention correctors being userobjects from times to times. What do you think?

GiudGiud replied:
I think it's fine. If users can't understand inheritance between Correctors and User Objects they won't like MOOSE much anyway. 
And developers can always use the reminder since the execution scheduling of Correctors is tied with the user objects, same as the postprocessors. So they do need to know

hugary1995 replied:
Yeah, I think we don't require users to have OOP background, so maybe we won't expect them to understand the specific word "inheritance", but I agree we can communicate to users that "correctors are user objects behind the scenes".  So I would suggest explaining this bit in `Correctors/index.md`.

GiudGiud commented:
I ll clean this up.
For future reference objects in modules are not to be moved in the framework casually

GiudGiud committed:
Add a new base for correctors to prevent regular UOs in Correctors/ syntax
Deprecate adding correctors in user object block
Remove deprecation from flow physics

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/2378110/) on 02e6182 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/28338/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/2378156/) on 02e6182 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>e0de65</th>
      <th colspan=4>#28338 02e618</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.93%</td>
      <td>85.02%</td>
      <td>+0.09%</td>
      <td>78.75%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>104783</td>
      <td>104914</td>
      <td>+131</td>
      <td>63</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>18593</td>
      <td>18480</td>
      <td>-113</td>
      <td>17</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28338/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28338/coverage/framework)


## Modules coverage

### Navier stokes

<table>
  <thead>
    <tr>
      <th></th>
      <th>e0de65</th>
      <th colspan=4>#28338 02e618</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.57%</td>
      <td>84.57%</td>
      <td>+0.00%</td>
      <td>96.61%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>16349</td>
      <td>16352</td>
      <td>+3</td>
      <td>57</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>2984</td>
      <td>2984</td>
      <td>-</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/28338/coverage_diff/navier_stokes)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/28338/coverage/navier_stokes)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/28338/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/28338/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/28338/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/28338/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/28338/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/28338/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/28338/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/28338/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/28338/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/28338/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/28338/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/28338/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/28338/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/28338/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/28338/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/28338/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/28338/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/28338/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/28338/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/28338/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/28338/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/28338/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/28338/coverage/scalar_transport)
- [`solid_mechanics`](https://mooseframework.inl.gov/docs/PRs/28338/coverage/solid_mechanics)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/28338/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/28338/coverage/stochastic_tools)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/28338/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/28338/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 78.75% is less than the suggested 90.0%



This comment will be updated on new commits.

hugary1995 commented on modules/navier_stokes/src/correctors/NSPressurePin.C:
```diff
@@ -48,6 +48,7 @@ NSPressurePin::validParams()
       "pressure_average", "A postprocessor that computes the average of the pressure variable");
 
   params.addClassDescription("Pins the pressure after a solve");
+  params.registerBase("Corrector");
```
Just curious, is this now necessary for every corrector class?

GiudGiud replied:
until we make a base class and decide what goes in it, yes

hugary1995 reviewed (changes requested):
LGTM

GiudGiud commented on framework/include/actions/AddCorrectorAction.h:
```diff
@@ -0,0 +1,31 @@
+{
+public:
+  /**
+   * Class constructor
```
ugh this was copy-pasted. I ll find the source and remove it there too

GiudGiud replied:
done

GiudGiud committed:
Address review
remove bad docstring on validParams

Apply suggestions from code review

Apply suggestions from code review

GiudGiud commented on framework/include/actions/AddCorrectorAction.h:
```diff
@@ -0,0 +1,27 @@
+
+/**
+ * Action for creating Corrector, a UserObject
+ * designed for nonlinear variable solutions without constraints.
```
```suggestion
 * designed for modifying nonlinear variable solutions, for example outside of the nonlinear solve
```

hugary1995 commented:
This looks good to me otherwise. I'll approve when all tests pass.

GiudGiud commented on framework/doc/content/syntax/Correctors/index.md:
```diff
@@ -0,0 +1,14 @@
+
+The `Correctors` system is designed to modify the values of nonlinear variable solutions.
+This can be as part of a predictor-corrector time integration scheme for example.
+
```
```suggestion

!alert note
Correctors are [UserObjects](UserObjects/index.md) behind the scene. They simply have a dedicated role and syntax.
```

GiudGiud commented:
@hugary1995

hugary1995 reviewed (approved):
For future reference: Having to call `registerBase` in every corrector is suboptimal. @GiudGiud said it will no longer be needed once a common base class for all Correctors is added.

GiudGiud commented:
I've been thinking about it and the inheritance tree of UOs would force us to create multiple base classes. One under generalUO, and maybe one under ElementUO

So maybe instead of inserting base classes in between UOs and Correctors, we should have an interface/helper like parent class that changes the base and adds whichever other attributes.
